Fixes#1238 Skip feature added in login

ujjwalagrawal17 opened the pull request:
## Description

Fixes #1238

The following changes are made to implement skip login feature:-
1. Added skip button in login.
2. Implemented onClick in skip button which redirects to NearbyImagesActivity
3. Added new sharedPreference parameter (login_skipped) - True if login is skipped else false.
4. Made navigation drawer  dynamic according to login_skipped parameter (checked via sharedprefs)

## Tests performed

Manually Tested on API 25 (MotoG5s+) with build variant, e.g. ProdDebug}.
1. Tested drawer by logging in, logout and by skipping, log.

## Screenshots showing what changed
 <table>
<tr>
<td>

![screenshot_20180502-194244](https://user-images.githubusercontent.com/19607555/39537334-dc0b87d4-4e56-11e8-909f-4391638fab7b.png)
</td>
<td>

![screenshot_20180502-194215](https://user-images.githubusercontent.com/19607555/39537335-dc899e30-4e56-11e8-91e7-e5f211a0fe1e.png)

</td>
</tr>
<tr>
<td>

![screenshot_20180502-194211](https://user-images.githubusercontent.com/19607555/39537338-dd02c6ca-4e56-11e8-82ee-06a635eab07d.png)

</td>
<td>

![screenshot_20180502-194332](https://user-images.githubusercontent.com/19607555/39537333-db60b8f4-4e56-11e8-8ed5-5c9b9ff83c8d.png)

</td>
</tr>
</table>

ujjwalagrawal17 commented:
Using this PR we will not be able to upload photos but we can integrate browse, nearby features.

ujjwalagrawal17 commented:
@neslihanturan @nicolas-raoul Can you review it.

neslihanturan commented:
Very good job @ujjwalagrawal17 , but requires discussion. I will do the testing, beside tests, team do you agree we should include this feature?

ujjwalagrawal17 commented:
@maskaravivek @misaochan  Should we add this feature?

nicolas-raoul commented:
What happens when someone not logged-in shares a picture from the Android gallery to the Commons app?

ujjwalagrawal17 commented:
Ohh. I missed it. I will add a check for that and if user is not logged in we will give a non-cancellable pop-up to login or goto nearby/featured image Activity. Should we do anything else @nicolas-raoul ?

misaochan commented:
Off the top of my head, things that we need to check for:

- Notifications cannot be accessed when not logged in
- The check needs to be done for camera and gallery, both from Nearby and from Contributions

ujjwalagrawal17 commented:
@misaochan Notification is already invisible so it cannot be accessed when not logged in. Also since user is not logged in we should not show contributions. And in nearby, add fab should be invisible in case of not logged in user,  I will do it.

misaochan commented:
@ujjwalagrawal17 Thanks for the clarification. It would be worth noting that in addition to making Nearby map's FAB invisible, Nearby list needs to have camera and gallery buttons greyed out (and not accessible) as well.

sivaraam commented:
Just a little suggestion, I guess it would better to avoid showing the log-in screen completely and then allow users to log-in only when they wish to use features that require login.

I suggest this way because, it's quite possible users wouldn't notice the ability to "Skip" the log-in. If a lot of people do miss it out then it's quite possible the amount of users who uninstall the app immediately after installing it wouldn't be reduced much (which I suppose is the one of the main goals of this PR).

Not showing the Login screen at all would have more impact, I think. Of course this would make more sense only if the user is able to do at least something useful without logging in (like browsing).

sivaraam commented:
In general, it would be nice to clearly communicate to the user what features aren't available when the user is not logged in.

nicolas-raoul commented:
@sivaraam Good idea, could you please create a new issue for it?

sivaraam commented:
> Good idea, could you please create a new issue for it?

No problem :-) Just for the sake of clarity which of the following comments are referring to:

1. [Comment 1](https://github.com/commons-app/apps-android-commons/pull/1486#issuecomment-399758989)
2. [Comment 2](https://github.com/commons-app/apps-android-commons/pull/1486#issuecomment-399758936)

nicolas-raoul commented:
@sivaraam I was referring to comment 1, sorry for the ambiguity.
Actually, rather than a new issue I added this comment: https://github.com/commons-app/apps-android-commons/issues/1238#issuecomment-399892860

sivaraam commented:
@nicolas-raoul The links for the "Comment 1" and "Comment 2" point differently in the [above comment](https://github.com/commons-app/apps-android-commons/pull/1486#issuecomment-399834645). Never mind, thanks for the clarification :relaxed:

ujjwalagrawal17 commented:
@sivaraam Nice suggestion. I am working on Category details page now. Will work on this feature after completing it.

neslihanturan commented:
@ujjwalagrawal17 as far as I remembered from todays meeting you said I should close this PR, and you will send a new one (or reopen existing one) after you are done with browse feature. Am I right?

ujjwalagrawal17 commented:
Umm, it would be helpful for me to work on same PR. can we keep this on hold for sometime @neslihanturan.

ujjwalagrawal17 commented:
@neslihanturan @nicolas-raoul I am continuing working on this now.

neslihanturan commented:
Don't forget to solve conflicts when you are done Ujjwall:)

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/auth/LoginActivity.java:
```diff
@@ -135,6 +147,13 @@ public void onCreate(Bundle savedInstanceState) {
         }
     }
 
+    private void skipLogin() {
```
Javadoc would be useful here

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/nearby/NearbyMapFragment.java:
```diff
@@ -490,6 +514,18 @@ public void onMapReady(MapboxMap mapboxMap) {
         mapView.setStyleUrl("asset://mapstyle.json");
     }
 
+    private class BaseLogoutListener implements CommonsApplication.LogoutListener {
```
javadoc

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -246,6 +246,11 @@
+  <string name="skip_login">Skip</string>
+  <string name="navigation_item_login">Login</string>
+  <string name="skip_login_title">Do you really want to skip login ?</string>
+  <string name="skip_login_message">You might not be able to access some features of the app.</string>
```
How about saying clearly "You will not be able to upload pictures."?

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -246,6 +246,11 @@
   <string name="view_browser">View in Browser</string>
+  <string name="skip_login">Skip</string>
+  <string name="navigation_item_login">Login</string>
+  <string name="skip_login_title">Do you really want to skip login ?</string>
```
Not space before `?` in English. In French there must be a space :-)

nicolas-raoul commented on app/src/main/res/values/strings.xml:
```diff
@@ -246,6 +246,11 @@
+  <string name="navigation_item_login">Login</string>
+  <string name="skip_login_title">Do you really want to skip login ?</string>
+  <string name="skip_login_message">You might not be able to access some features of the app.</string>
+  <string name="login_alert_message">This feature requires user to be logged in !!</string>
```
No space before `!`, and I think a single `!` is enough :-)

ujjwalagrawal17 commented:
@nicolas-raoul the PR is not yet completed. The featured images are not showing when user is not logged in so fixing that now. Also I will make the changes told by you.

ujjwalagrawal17 commented:
@nicolas-raoul @neslihanturan @misaochan @maskaravivek 
The order in which I planned skip feature should work was Welcome Screen will come in first run after that everytime explore activity will come if user is logged in or not. The drawer menu will changed if user is not logged in

sivaraam commented on app/src/main/res/values/strings.xml:
```diff
@@ -246,6 +246,11 @@
   <string name="nominated_see_more"><u>See webpage for details</u></string>
   <string name="view_browser">View in Browser</string>
+  <string name="skip_login">Skip</string>
+  <string name="navigation_item_login">Login</string>
```
Wait. I thought we agreed on [not showing the log-in screen at all](1486#issuecomment-399758936) instead including a "Skip" button in the log-in page. So, why do we still have the "Skip login" strings? Am I missing something?

**Note**: I haven't reviewed the changes *completely* (or more correctly, I might not understand what has been changed and how.) . I haven't tested app with these proposed changes.

nicolas-raoul replied:
That confused me a bit too. Indeed the "Skip" link could be removed, it is not useful, users who suddenly decide that after all they don't want to log in can just press their "Back" button.

misaochan commented:
Could I please request that upon completion of this PR, we get a brief report on the functions that were tested **with** and **without** logging in, and the results? This will help the tester decide on merging the PR. My requested tests:

1. In-app camera and gallery uploads, and Nearby uploads
2. "Share" from Android gallery
3. All the available items on the nav drawer
4. The Achievements button

ujjwalagrawal17 commented:
The PR is ready to review.

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/theme/NavigationBaseActivity.java:
```diff
@@ -71,7 +71,7 @@ public void initDrawer() {
         View headerLayout = navigationView.getHeaderView(0);
         ImageView userIcon = headerLayout.findViewById(R.id.user_icon);
-        if (prefs.getBoolean("login_skipped", true)) {
+        if (prefs.getBoolean("login_skipped", false)) {
```
Does "login_skipped=false" mean that login has been skipped?

ujjwalagrawal17 replied:
login_skipped ==true means login skipped.

ujjwalagrawal17 commented:
### **Screenshots: **
 **LoggedIn**

- In-app camera and gallery uploads, and Nearby uploads (also from camera and gallery) (working as before)
- "Share" from Android gallery (working as before)
- All the available items on the nav drawer (working as before)
- The Achievements button (working as before)

<table>
<tr>
<td>

![screenshot_20180731-205934](https://user-images.githubusercontent.com/19607555/43488271-bed2e3ca-9536-11e8-85f1-f27d9f572622.png)

</td>
<td>

![screenshot_20180731-205912](https://user-images.githubusercontent.com/19607555/43488329-f8d8438a-9536-11e8-9339-b54dedda3253.png)

</td>
<td>

![screenshot_20180731-205715](https://user-images.githubusercontent.com/19607555/43488330-f94274a8-9536-11e8-9e48-1c583a2e241c.png)

</td>
</tr>
</table>

**Login Skipped**
- In-app camera and gallery uploads, and Nearby uploads (also from camera and gallery) (Popup is shown in all conditions, contibutions is not visible so only nearby needs to be changed)
- "Share" from Android gallery (Sending user to login screen. Refer to screenshot) 
- All the available items on the nav drawer (Contributions, notifications are not visible) 
- The Achievements button (not visible as user is not logged In)


<table>
<tr>
<td>

![screenshot_20180731-210017](https://user-images.githubusercontent.com/19607555/43488327-f87334e0-9536-11e8-8910-437e2f090458.png)

</td>
<td>

![screenshot_20180801-025409](https://user-images.githubusercontent.com/19607555/43488269-bdeae41c-9536-11e8-90e0-33300fe98c98.png)

</td>
<td>

![screenshot_20180731-210109](https://user-images.githubusercontent.com/19607555/43488270-be5a2610-9536-11e8-939d-cc8393bca393.png)

</td>
</tr>
<tr>
<td>


</td>
<td>


</td>
<td>


</td>
</tr>
</table>

nicolas-raoul commented:
@sivaraam Good idea! Would you mind performing these tests? :-)

misaochan commented:
Thanks for the screenshots and tests, @ujjwalagrawal17 . :) They look good to me, great job. Perhaps for "You need to login first!" and "This feature requires user to login!", we can replace both of these strings with "Please log in to use this feature"?

Happy to merge pending successful tests from a third party as @sivaraam suggested.

nicolas-raoul commented:
No third party willing to test?
Can I be the "third party"? :-)

sivaraam commented:
> No third party willing to test?

I wouldn't mind testing it. But I'm busy till weekend.

> Can I be the "third party"? :-)

If this needs to be moved forward quickly and can't wait till weekend (when I get free), you can test it. I especially asked a review from a third person as I never saw the screen-shot of how the initial screen looks now. @ujjwalagrawal17  told he would design it as @misaochan [suggests in #1238](1238#issuecomment-408597610) but never posted a screen shot of how it looks ;-) Or I missed something?

nicolas-raoul commented:
Anyway, I did the testing too:

# When logged in

- in-app camera/gallery, share from android stock gallery, nearby camera/gallery: All work except upload fails, probably due to the other issue
- Achievements work
- Home/Nearby/Explore/Settings/About/Tutorial/Feedback/Logout work
- "Notifications" does not work: white screen

# When logged out

- No achievements.
- The "Log in" button has the achievements icon, why not.
- No in-app camera/gallery
- Nearby says "You can upload a picture for any place from your gallery or camera", I guess it is OK to explain the feature even if it is not usable right now. Pressing the + shows a popup saying to log in, correct.
- Share from gallery: The title/description activity pops up, and only when pressing "Send" does the log in prompt appears. After logging in I am shown "My recent Uploads". The title/description that I spent time writing seems to be lost :'-(

Another thing I noticed:
Pressing "Skip" takes a very long time, it feels very unresponsive, I always have to press it several times.

misaochan commented:
Thanks for the tests, @nicolas-raoul !

ujjwalagrawal17 commented:
- Changed strings according to @misaochan 's Comment
- Fixed share from gallery when not logged In it is directly taking the user to the login page instead of asking title, description.
- Skip button seems to work fine on my phone. (Similar to forgot password) @nicolas-raoul I can try increasing padding to work it more efficiently.

nicolas-raoul commented:
Trying it again, the skip login button now appears to work fast, every time. Strange, maybe it was a slow down of my phone for another reason. I will close the issue for now, and reopen if I find a way to reproduce.

ujjwalagrawal17 commented:
@nicolas-raoul  Can we merge this now

neslihanturan commented:
@nicolas-raoul do you still see white screen when you logged in?

> ⚠ "Notifications" does not work: white screen

nicolas-raoul commented:
In any case it is most probably unrelated so I think this can be merged.

nicolas-raoul commented:
Notifications seem to work now, it was probably a network problem or something, I created https://github.com/commons-app/apps-android-commons/issues/1782 for that problem.
